hugo/assets/js/species_cards.js

```diff
@@ -0,0 +1,284 @@
+        const species = {
+            id: template.id,
+            speciesText: title + ' ' + subtitle, // used for searching
+            last_updated: template.content.querySelector('.card-text').textContent.replace('Last updated: ', ''),
+        };
+        state.speciesData.push(species);
+    });
+    state.numbMatches = templates.length
+
+    renderSpeciesCards();
+}
+
+
+// Sorting functions below
+
+function stringToDate(str) {
+    // Need to use reformat date standard to use JS Date constructor
+    const [dd, mm, yyyy] = str.split('/');
+    return new Date(yyyy, mm - 1, dd);
+}
+
+function sortLastUpdated(a, b) {
+    const dateA = stringToDate(a.last_updated);
+    const dateB = stringToDate(b.last_updated);
+    return dateB - dateA;
+}
+
+function sortAlphabetically(a, b) {
+    return a.speciesText.localeCompare(b.speciesText);
+}
+
+const sortRevAlphabetically = (a, b) => -sortAlphabetically(a, b);
+
+
+/**
+ * Run on page load and every time an event listener is triggered.
+ * Based on current state, updates: species cards, pagination buttons and the no results card.
+ */
+function renderSpeciesCards() {
+    const cardContainer = document.getElementById('card-container');
+    cardContainer.innerHTML = '';
+
+    const speciesToDisplay = filterAndOrderSpecies();
+
+    if (speciesToDisplay.length === 0) {
+        noResultsCard.style.display = 'block';
+        state.currentPage = 1
+        updatePaginationButtons();
+    } else {
+        noResultsCard.style.display = 'none';
+        speciesToDisplay.forEach((speciesId) => {
+            const template = document.getElementById(speciesId);
+            const content = template.content.cloneNode(true);
+            cardContainer.appendChild(content);
+        });
+        updatePaginationButtons();
+    }
+}
+
+
+/**
+ * Search/filter all species and order them according to active dropdown selection
+ * The id's of each species <template> is returned (in order of display).
+ *
+ * @returns {Array} - An array of filtered species HTML <template> ids.
+ *                    (Returns empty array if no matches.)
+ */
+function filterAndOrderSpecies() {
+    let filteredSpecies = state.speciesData;
+
+    // filter by text content
+    if (state.searchText !== "") {
+        filteredSpecies = state.speciesData.filter(species => {
+            return species.speciesText.includes(state.searchText);
+        });
+    }
+
+    state.numbMatches = filteredSpecies.length
+
+    // order species
+    if (filteredSpecies.length !== 0) {
+        if (state.sortOrder === SORT_OPTIONS.LAST_UPDATED) {
+            filteredSpecies.sort(sortLastUpdated);
+        } else if (state.sortOrder === SORT_OPTIONS.A_TO_Z) {
+            filteredSpecies.sort(sortAlphabetically);
+        } else {
+            filteredSpecies.sort(sortRevAlphabetically);
+        }
+    }
+
+    // select results to show based on page number
+    if (PAGINATION_EXISTS) {
```

Comment:
You could separate filtering and pagination by introducing a `getCurrentPage(speciesList, state)` function that could be ...you guessed it, tested separately :)

Reply:
Yes! This happened naturally once I changed to limit where I update the state object. Nice to follow the single responsibility principle as much as possible too :)